drivers/wifi/infineon/cyw43xxx_drv.c

```diff
@@ -0,0 +1,396 @@
+
+	if (add_remove_amount > 0) {
+		(*buf)->len = (*buf)->size;
+		(*buf)->data = net_buf_pull(*buf, add_remove_amount);
+	} else {
+		(*buf)->data = net_buf_push(*buf, -add_remove_amount);
+		(*buf)->len = (*buf)->size;
+	}
+	return WHD_SUCCESS;
+}
+
+static int eth_infineon_cyw43xxx_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct infineon_cyw43xxx_wifi_runtime *data = dev->data;
+	cy_rslt_t ret = CY_RSLT_SUCCESS;
+	size_t pkt_len = net_pkt_get_len(pkt);
+	struct net_buf *buf = NULL;
+
+	/* Read the packet payload */
+	if (net_pkt_read(pkt, data->frame_buf, pkt_len) < 0) {
+		LOG_ERR("net_pkt_read failed");
+		return -EIO;
+	}
+	/* Allocate Network Buffer from pool with Packet Length + Data Header */
+	buf = net_buf_alloc_len(&cyw43xxx_pool, pkt_len + sizeof(data_header_t), K_NO_WAIT);
+
+	/* Reserve the buffer Headroom for WHD Data header */
+	net_buf_reserve(buf, sizeof(data_header_t));
+
+	/* Copy the buffer to network Buffer pointer */
+	(void)memcpy(buf->data, data->frame_buf, pkt_len);
+
+	/* Call WHD API to send out the Packet */
+	ret = whd_network_send_ethernet_data(cyw43xxx_if, (void *)buf);
+	if (ret != CY_RSLT_SUCCESS) {
+		LOG_ERR("whd_network_send_ethernet_data failed");
+		return -EIO;
+	}
+	return 0;
+}
+
+void cy_network_process_ethernet_data(whd_interface_t interface, whd_buffer_t buffer)
+{
+	struct net_pkt *pkt;
+	uint8_t *data = whd_buffer_get_current_piece_data_pointer(interface->whd_driver, buffer);
+	uint32_t len = whd_buffer_get_current_piece_size(interface->whd_driver, buffer);
+	bool net_pkt_unref_flag = false;
+
+	if (infineon_cyw43xxx_wifi_iface == NULL) {
+		LOG_ERR("network interface unavailable");
+		cy_buffer_release(buffer, WHD_NETWORK_RX);
+		return;
+	}
+
+	pkt = net_pkt_rx_alloc_with_buffer(infineon_cyw43xxx_wifi_iface, len,
+					   AF_UNSPEC, 0, K_NO_WAIT);
+
+	if (pkt != NULL) {
+		if (net_pkt_write(pkt, data, len) < 0) {
+			LOG_ERR("Failed to write pkt");
+			net_pkt_unref_flag = true;
+		}
+
+		if ((net_pkt_unref_flag) ||
+		    (net_recv_data(infineon_cyw43xxx_wifi_iface, pkt) < 0)) {
+			LOG_ERR("Failed to push received data");
+			net_pkt_unref_flag = true;
+		}
+	} else {
+		LOG_ERR("Failed to get net buffer");
+	}
+
+	/* Release a packet buffer */
+	cy_buffer_release(buffer, WHD_NETWORK_RX);
+
+	if (net_pkt_unref_flag) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void *link_events_handler(whd_interface_t ifp,
+				 const whd_event_header_t *event_header,
+				 const uint8_t *event_data, void
+				 *handler_user_data)
+{
+	ARG_UNUSED(ifp);
+	ARG_UNUSED(event_data);
+	ARG_UNUSED(handler_user_data);
+
+	(void)k_msgq_put(&cyw43xxx_wifi_msgq, event_header, K_FOREVER);
+	return NULL;
+}
+
+static void infineon_cyw43xxx_event_task(void)
+{
+	whd_event_header_t event_header;
+
+	while (1) {
+		(void)k_msgq_get(&cyw43xxx_wifi_msgq, &event_header, K_FOREVER);
+
+		switch ((whd_event_num_t)event_header.event_type) {
+		case WLC_E_LINK:
+			(void)net_if_up(infineon_cyw43xxx_wifi_iface);
```

Comment:
Networking drivers shouldn't be calling `net_if_up` and `net_if_down` themselves.
https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html#network-interface-state-management
Network association should be done with the dormant state

Reply:
hi @JordanYates,

if we talk about sample/net/wifi is common, ` net iface up 1` need to execute for user before perform net operation (e.g. `net ping ....`). Is it correct understanding?